subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,743 @@
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	LOG_DBG("Get Protocol: %s", ddata->protocol ? "Report" : "Boot");
+	net_buf_add_u8(buf, ddata->protocol);
+
+	return 0;
+}
+
+static int handle_get_descriptor(const struct device *dev,
+				 const struct usb_setup_packet *const setup,
+				 struct net_buf *const buf)
+{
+	struct hid_device_data *const ddata = dev->data;
+	uint8_t desc_type = USB_GET_DESCRIPTOR_TYPE(setup->wValue);
+	uint8_t desc_idx = USB_GET_DESCRIPTOR_INDEX(setup->wValue);
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	switch (desc_type) {
+	case USB_DESC_HID_REPORT:
+		LOG_DBG("Get descriptor report");
+		net_buf_add_mem(buf, ddata->rdesc, MIN(ddata->rsize, setup->wLength));
+		break;
+	case USB_DESC_HID:
+		LOG_DBG("Get descriptor HID");
+		net_buf_add_mem(buf, &desc->hid, MIN(desc->hid.bLength, setup->wLength));
+		break;
+	case USB_DESC_HID_PHYSICAL:
+		LOG_DBG("Get descriptor physical %u", desc_idx);
+		errno = -ENOTSUP;
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return 0;
+}
+
+static int usbd_hid_ctd(struct usbd_class_data *const c_data,
+			const struct usb_setup_packet *const setup,
+			const struct net_buf *const buf)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	int ret = 0;
+
+	switch (setup->bRequest) {
+	case USB_HID_SET_IDLE:
+		ret = handle_set_idle(dev, setup);
+		break;
+	case USB_HID_SET_REPORT:
+		ret = handle_set_report(dev, setup, buf);
+		break;
+	case USB_HID_SET_PROTOCOL:
+		ret = handle_set_protocol(dev, setup);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+
+static int usbd_hid_cth(struct usbd_class_data *const c_data,
+			const struct usb_setup_packet *const setup,
+			struct net_buf *const buf)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	int ret = 0;
+
+	switch (setup->bRequest) {
+	case USB_HID_GET_IDLE:
+		ret = handle_get_idle(dev, setup, buf);
+		break;
+	case USB_HID_GET_REPORT:
+		ret = handle_get_report(dev, setup, buf);
+		break;
+	case USB_HID_GET_PROTOCOL:
+		ret = handle_get_protocol(dev, setup, buf);
+		break;
+	case USB_SREQ_GET_DESCRIPTOR:
+		ret = handle_get_descriptor(dev, setup, buf);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+
+static void usbd_hid_sof(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *const ops = ddata->ops;
+
+	if (ops->sof) {
+		ops->sof(dev);
+	}
+}
+
+static void usbd_hid_enable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *const ops = ddata->ops;
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
+	ddata->protocol = HID_PROTOCOL_REPORT;
+	if (ops->iface_ready) {
+		ops->iface_ready(dev, true);
```

Comment:
You should not mix them, HID device is one part, whatever else you have related to USB device management is another. iface_ready() is there for convenience, even with a message type like "configured:num" you would need to check, depending on the USB configuration, if it applies to the configuration and speed the HID device instance belongs to or not.

Reply:
I understand that these are 2 separate things.

Still, maybe we could think about ensuring `iface_ready` notification order in relation to `UDC_EVT_`. It may be a bit confusing for an application if it e.g. gets `iface_ready` before `USBD_MSG_VBUS_READY` (e.g. because of blocked system workqueue). Or if resume from suspend is delayed for noticeable time after `iface_ready` was already called. Maybe we could e.g. keep the `iface_ready` API as is and just call application's callback through the system workqueue to ensure proper order?

Reply:
No. Again, HID device iface_ready is low level, for the users top half of the driver. Do not mix them. Suspended bus does not mean that HID device is not ready.

Reply:
It is perfectly fine for host to suspend HID device that supports Remote Wakeup for aggressive power saving. It is up to device to initiate the Remote Wakeup. It makes sense for device to initiate wakeup only if it has something to send to host and therefore suspend/resume should not affect `iface_ready`.

Reply:
Thanks for clarification